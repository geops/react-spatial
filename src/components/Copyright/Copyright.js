import React, { Component } from 'react';
import PropTypes from 'prop-types';
import LayerService from '../../LayerService';

const propTypes = {
  /**
   * Layer Service.
   */
  layerService: PropTypes.instanceOf(LayerService).isRequired,

  /**
   * Format function. Called with an array of copyrights from visible layers
   * and returns the copyright.
   */
  format: PropTypes.func,
};

const defaultProps = {
  format: copyrights => (
    <>
      &copy;
      {` ${copyrights.join(' | ')}`}
    </>
  ),
};

class Copyright extends Component {
  constructor(props) {
    super(props);
    this.state = {
      copyrights: [],
    };
  }

  componentDidMount() {
    this.updateLayerService();
  }

  componentDidUpdate(prevProps) {
    const { layerService } = this.props;
    if (layerService !== prevProps.layerService) {
      this.updateLayerService();
    }
  }

  componentWillUnmount() {
    const { layerService } = this.props;
    layerService.unlistenChangeEvt('change:visible');
  }

  updateLayerService() {
    const { layerService } = this.props;
    layerService.on('change:visible', () => this.updateCopyright());
    this.updateCopyright();
  }

  updateCopyright() {
    const { layerService } = this.props;

    const copyrights = layerService
      .getLayersAsFlatArray()
      .filter(l => l.getVisible() && l.getCopyright())
      .map(l => l.getCopyright());

    // remove duplicates
    const unique = Array.from(new Set(copyrights));

    this.setState({
      copyrights: unique,
    });
  }

  render() {
    const { format } = this.props;
    const { copyrights } = this.state;

    if (!copyrights.length) {
      return null;
    }

    return <div className="tm-copyright">{format(copyrights)}</div>;
  }
}

Copyright.propTypes = propTypes;
Copyright.defaultProps = defaultProps;
export default Copyright;
