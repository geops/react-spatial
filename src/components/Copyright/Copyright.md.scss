.rs-copyright-example {
  .rs-copyright {
    font-size: 11px;
  }
}
