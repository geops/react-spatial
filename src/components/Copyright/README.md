#

This demonstrates the use of Copyright.

```js
import React, { Component } from  'react';
import BasicMap from 'react-spatial/components/BasicMap';
import ConfigReader from 'react-spatial/ConfigReader';
import LayerService from 'react-spatial/LayerService';
import OLMap from 'ol/Map';
import Footer from 'react-spatial/components/Footer';
import Copyright from './Copyright';

class CopyrightExample extends Component {
  constructor(props) {
    const layerConf = [{
      name: 'OSM Baselayer',
      visible: true,
      copyright: 'OSM Contributors',
      data: {
        type: 'xyz',
        url: 'https://c.tile.openstreetmap.org/{z}/{x}/{y}.png',
      },
    }];

    this.map = new OLMap();
    this.layers = ConfigReader.readConfig(this.map, layerConf);
    this.layerService = new LayerService(this.layers);
  }

  render() {
    return (
      <div className="tm-scale-line-example">
        <BasicMap
          map={this.map}
          zoom={3}
          layers={this.layers}
        />
        <Footer>
          <Copyright layerService={this.layerService} />
        </Footer>
      </div>
    );
  }
}

<CopyrightExample />
```
