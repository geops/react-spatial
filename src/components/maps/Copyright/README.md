#

This demonstrates the use of Copyright.

```js
import React from  'react';
import ConfigReader from 'react-spatial/ConfigReader';
import LayerService from 'react-spatial/LayerService';
import Copyright from 'react-spatial/components/maps/Copyright';

const layerConf = [{
  name: 'OSM Baselayer',
  visible: true,
  copyright: '&copy; OSM Contributors',
  data: {
    type: 'xyz',
    url: 'https://c.tile.openstreetmap.org/{z}/{x}/{y}.png',
  },
}];
const layerService = new LayerService(ConfigReader.readConfig(layerConf));

<Copyright layerService={layerService} />
```
