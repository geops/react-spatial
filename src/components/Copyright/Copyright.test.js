import "jest-canvas-mock";
import React from "react";
import { Map, View } from "ol";
import { render, act } from "@testing-library/react";
import Tile from "ol/Tile";
import TileLayer from "ol/layer/Tile";
import TileSource from "ol/source/Tile";
import { createXYZ } from "ol/tilegrid";
import Copyright from "./Copyright";

const image = new Image();
image.width = 256;
image.height = 256;

const tileLoadFunction = () => {
  const tile = new Tile([0, 0, -1], 2 /* LOADED */);
  tile.getImage = () => {
    return image;
  };
  return tile;
};

const getOLTileLayer = (options = {}) => {
  const layer = new TileLayer({
    ...options,
    source: new TileSource({
      projection: "EPSG:3857",
      tileGrid: createXYZ(),
    }),
  });
  layer.getSource().getTile = tileLoadFunction;
  return layer;
};

const getLayer = (copyrights, visible = true) => {
  return getOLTileLayer({
    visible,
    copyrights,
  });
};

let layers;
let map;

describe("Copyright", () => {
  beforeEach(() => {
    const target = document.createElement("div");
    document.body.appendChild(target);
    layers = [getLayer("bar"), getLayer("foo", false)];
    map = new Map({
      target,
      view: new View({
        center: [0, 0],
        zoom: 0,
      }),
    });
    map.setSize([200, 200]);
    layers.forEach((layer) => {
      map.addLayer(layer);
    });
    act(() => {
      map.renderSync();
    });
  });

  afterEach(() => {
    layers.forEach((layer) => {
      map.removeLayer(layer);
    });
    map.setTarget(null);
    map = null;
  });

  test("is empty if no layers are visible", () => {
    const { container } = render(<Copyright map={map} />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test("displays one copyright", () => {
    const { container } = render(<Copyright map={map} />);
    act(() => {
      map.renderSync();
    });
    expect(container.textContent).toBe("bar");
  });

  test("displays 2 copyrights", () => {
    const { container } = render(<Copyright map={map} />);
    layers[0].setVisible(true);
    layers[1].setVisible(true);
    act(() => {
      map.renderSync();
    });
    act(() => {
      map.renderSync();
    });
    expect(container.textContent).toBe("bar | foo");
  });

  test("displays a copyright using a custom format", () => {
    const { container } = render(
      <Copyright
        map={map}
        format={(copyrights) => {
          return `Number of copyrights: ${copyrights.length}`;
        }}
      />,
    );

    act(() => {
      map.renderSync();
    });

    expect(container.textContent).toBe("Number of copyrights: 1");
  });

  test("set a custom className", () => {
    const { container } = render(<Copyright map={map} className="foo" />);

    act(() => {
      map.renderSync();
    });

    expect(container.querySelectorAll(".foo").length).toBe(1);
  });

  test("set a custom attribute to the root element", () => {
    const { container } = render(
      <Copyright map={map} className="lala" foo="bar" />,
    );

    act(() => {
      map.renderSync();
    });

    expect(container.querySelectorAll("[foo]").length).toBe(1);
  });
});
