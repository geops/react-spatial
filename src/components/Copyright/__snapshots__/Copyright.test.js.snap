// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Copyright is empty if no layers are visible 1`] = `
<div class="rs-copyright">
  <div>
  </div>
</div>
`;
